### DISCUSSION TITLE
(testecho.cpp) How does client specify which server to connect to?

### DISCUSSION THREAD

**Discussion by hy2850** [ANLAB-KAIST/KENSv3#34] [Project Q&A 1. Application Layer (Socket)]
TL;DR : In testecho.cpp, each server instance seems to be listening at '0.0.0.0'. So if there are multiple server instances running at the same time, how does client choose which server to connect to?

We have a 'many-to-many' test case where many servers and clients are running at the same time.
`SetUp` function in `testecho.cpp/EchoTesting` sets the server instances listening at '0.0.0.0', allowing every connection on local machine ([ref](https://serverfault.com/questions/78048/whats-the-difference-between-ip-address-0-0-0-0-and-127-0-0-1)).

https://github.com/ANLAB-KAIST/KENSv3/blob/f7d1641721fcd644750a61309e171faf5d8d4e7e/app/echo/testecho.cpp#L130-L133

On the other hand, client instance tries to connect to the IPv4 address '10.0.0.server_id'.

https://github.com/ANLAB-KAIST/KENSv3/blob/f7d1641721fcd644750a61309e171faf5d8d4e7e/app/echo/testecho.cpp#L144-L148

Then, how does each client choose which server to connect to, when there are multiple server instances listening to '0.0.0.0', and client is trying to connect to '10.0.0.server_id'?

Although understanding `testecho.cpp` is not necessary in PA1, I'm curious how the test code works. Please elaborate me if I have missed something.

Thank you for your time.

**Comment by Pusnow** [comment]
This question looks like how KENS works.

KENS is a network simulator, and it does not use *real* networking stack in the kernel.
If you call network system calls in the `serverMain` and `clientMain`, the requests are passed to KENS's TCP networking stack, and the stack simulates network environments.

https://github.com/ANLAB-KAIST/KENSv3/blob/6131eca47a7a9c069155b97386b19dc7fd099358/src/Networking/TCP/E_TCPApplication.cpp#L14-L104

If you want to know more, see KENS's [paper](https://dl.acm.org/doi/10.1145/2676723.2677304).
